spokestack/nlu/tflite.py

```diff
@@ -0,0 +1,119 @@
+        self._intent_meta = {
+            intent.pop("name"): intent for intent in self._metadata["intents"]
+        }
+        self._warm_up()
+
+    def __call__(self, utterance: str) -> Dict[str, Any]:
+        """ Forward Pass
+
+        Args:
+            utterance (str): string that needs to be understood
+
+        Returns: intents, slots, and model confidence
+
+        """
+        inputs, input_ids = self._encode(utterance)
+        outputs = self._model(inputs)
+        intent, tags, confidence = self._decode(outputs)
+
+        # slice off special tokens: [CLS], [SEP]
+        tags = tags[: len(input_ids) - 2]
+        input_ids = input_ids[1:-1]
+
+        # retrieve slots from the tagged postions and decode slots back
+        # into original values
+        slots = [token_id for token_id, tag in zip(input_ids, tags) if tag != "o"]
+        slots = self._tokenizer.decode(slots)
+
+        # attempt to resolve tagged tokens into slots and
+        # collect the successful ones
+        slot_meta = self._intent_meta[intent]["slots"]
+        parsed_slots = []
+        for meta in slot_meta:
+            parsed_slots.append(self._parse_slots(meta, slots))
+
+        return {
+            "utterance": utterance,
+            "intent": intent,
+            "confidence": confidence,
+            "slots": parsed_slots,
+        }
+
+    def _warm_up(self) -> None:
+        # make an array the same size as the inputs to warm the
+        # model since first inference is always slower than subsequent
+        warm = np.zeros((self._model.input_details[0]["shape"]), dtype=np.int32)
+        _ = self._model(warm)
+
+    def _encode(self, utterance: str) -> Tuple[np.ndarray, List[int]]:
+        inputs = self._tokenizer.encode(utterance)
+        # get the non-padded/truncated token ids to match the
+        # original utterance to the respective labels and
+        # use the length to slice the results
+        input_ids = inputs.ids
+        # it's (max_length + 1) because the [CLS]
+        # token gets appended inside the model
+        # notice the slice [1:] when we convert to an array
+        inputs.truncate(max_length=self._max_length + 1)
+        inputs.pad(length=self._max_length + 1)
+        inputs = np.array(inputs.ids[1:], np.int32)
+        # add the batch dimension for the TFLite model
+        inputs = np.expand_dims(inputs, 0)
+        return inputs, input_ids
+
+    def _decode(self, outputs) -> Tuple[str, List[str], float]:
+        # to get the index of the highest probability we
+        # apply argmax to the posteriors which allows the
+        # labels to be decoded with an integer to string mapping
+        # we derive the confidence from the highest probability
+        intent_posterior, tag_posterior = outputs
+        intents = self._decode_intent(intent_posterior)
+        tags = self._decode_tags(tag_posterior)
+        confidence = np.max(intent_posterior)
```

Comment:
This extra trip through the posteriors wouldn't be necessary if you let `_decode_intent` return both the index and value of the argmax.